tree/ntuple/v7/inc/ROOT/RPagePool.hxx

```diff
@@ -33,48 +33,104 @@ namespace Internal {
+   /// Adds a new page to the pool. Upon registration, the page pool takes ownership of the page's memory.
+   /// The new page has its reference counter set to 1.
+   RPageRef RegisterPage(RPage page);
    /// Like RegisterPage() but the reference counter is initialized to 0
-   void PreloadPage(const RPage &page, const RPageDeleter &deleter);
+   void PreloadPage(RPage page);
    /// Tries to find the page corresponding to column and index in the cache. If the page is found, its reference
    /// counter is increased
-   RPage GetPage(ColumnId_t columnId, NTupleSize_t globalIndex);
-   RPage GetPage(ColumnId_t columnId, RClusterIndex clusterIndex);
-   /// Give back a page to the pool and decrease the reference counter. There must not be any pointers anymore into
-   /// this page. If the reference counter drops to zero, the page pool might decide to call the deleter given in
-   /// during registration.
-   void ReturnPage(const RPage &page);
+   RPageRef GetPage(ColumnId_t columnId, NTupleSize_t globalIndex);
+   RPageRef GetPage(ColumnId_t columnId, RClusterIndex clusterIndex);
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Internal::RPageRef
+\ingroup NTuple
+\brief Reference to a page stored in the page pool
+
+The referenced page knows about its page pool and decreases the reference counter on destruction.
+*/
+// clang-format on
+class RPageRef {
+   friend class RPagePool;
+
+   RPage fPage;
+   RPagePool *fPagePool = nullptr;
+
+   // Called as delegated constructor and directly by the page pool
+   RPageRef(const RPage &page, RPagePool *pagePool) : fPagePool(pagePool)
+   {
+      // We leave the fPage::fPageAllocator member unset (nullptr), since fPage is a non-owning view on the page
+      fPage.fColumnId = page.fColumnId;
+      fPage.fBuffer = page.fBuffer;
+      fPage.fElementSize = page.fElementSize;
+      fPage.fNElements = page.fNElements;
+      fPage.fMaxElements = page.fMaxElements;
+      fPage.fRangeFirst = page.fRangeFirst;
+      fPage.fClusterInfo = page.fClusterInfo;
+   }
+
+public:
+   RPageRef() = default;
+   RPageRef(const RPageRef &other) = delete;
+   RPageRef &operator=(const RPageRef &other) = delete;
+
+   RPageRef(RPageRef &&other) : RPageRef(other.fPage, other.fPagePool) { other.fPagePool = nullptr; }
+
+   RPageRef &operator=(RPageRef &&other)
+   {
+      if (this != &other) {
+         std::swap(fPage, other.fPage);
+         std::swap(fPagePool, other.fPagePool);
+      }
+      return *this;
+   }
+
+   ~RPageRef()
+   {
+      if (fPagePool)
+         fPagePool->ReturnPage(fPage);
+   }
+
+   /// Used by the friend virtual page source to map the physical column and cluster IDs to ther virtual counterparts
+   void ChangeIds(DescriptorId_t columnId, DescriptorId_t clusterId)
+   {
+      fPage.fColumnId = columnId;
+      fPage.fClusterInfo = RPage::RClusterInfo(clusterId, fPage.fClusterInfo.GetIndexOffset());
+   }
+
+   const RPage &Get() const { return fPage; }
```

Comment:
Given how often it's used, I would consider adding an `operator->` here. Although it might obfuscate the code and not be very consistent with a "reference"-type access (e.g. `reference_wrapper`), so I'm not really sure about it.

Reply:
I was considering it but eventually decided against the `->` operator for the reason you mention.